File: 0322-coin-change/0322-coin-change.cpp
/*
//@Author: KING-SEN
// Programming Language Used: C++

class Solution {
public:
    int solve(vector<int>& coins, int amount, vector<int>& dp){
        if(amount < 0) return INT_MAX / 2;
        if(amount == 0) return 0;
        
        int& ret = dp[amount];
        
        if(ret != -1) return ret;
        ret = INT_MAX / 2;
        
        for(auto coin : coins){
            ret = min(ret, 1 + solve(coins, amount - coin, dp));
        }
        
        return ret;
    }
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount + 1, -1);
        int answer = solve(coins, amount, dp);
        
        if(answer >= INT_MAX / 2) return -1;
        return answer;
    }
};

// Time Complexity: O(Amount * N), N = denomination count.In the worst case the recursive tree of the algorithm has of height S and the algorithm solves only Amount subproblems because  it caches precalculated solutions in a table. Each subproblem is computed with N iterations. Therefore there is O(Amount * n) time complexity
// Space Complexity:  O(Amount), where we use extra space for the memoization table for amount.
*/
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int dp[amount + 1];
        for(int i = 0; i<=amount; i++) dp[i] = 1e9;
        dp[0] = 0;
        for(int i = 1; i<=amount; i++){
            for(int j=0; j<coins.size(); j++){
                if(coins[j] <= i) dp[i] = min(dp[i], dp[i - coins[j]] + 1);
            }
        }
        return dp[amount] == 1e9 ? -1 : dp[amount];
    }
};

File: 0322-coin-change/0322-coin-change.py
class Solution:
    # def coinChange(self, coins: List[int], amount: int) -> int:
    #     dp = defaultdict(lambda: -1)
    #     def solve(amount):
    #         if amount < 0:
    #             return float("inf")
    #         if amount == 0:
    #             return 0
    #         if dp[amount] != -1:
    #             return dp[amount]
    #         res = float("inf")
    #         for coin in coins:
    #             res = min(res, 1 + solve(amount - coin))
    #         dp[amount] = res
    #         return res
    #     ans = solve(amount)
    #     return -1 if ans >= float("inf") else ans
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [float("inf")] * (amount + 1)
        dp[0] = 0
        for i in range(amount + 1):
            for c in coins:
                if c <= i:
                    dp[i] = min(dp[i], 1 + dp[i - c])
        return dp[amount] if dp[amount] != float('inf') else -1 
            